### PR TITLE
FEAT: Add `is_dominated`

### LINKED ISSUE #44: Add is_dominated
QuantEcon/QuantEcon.py#327

Use [MathProgBase.jl](https://github.com/JuliaOpt/MathProgBase.jl).

### PR DESCRIPTION
**Pull request by QBatista**
Close #44

### CODE REVIEW

File: src/normal_form_game.jl [hunk 1]
```diff
@@ -691,3 +691,77 @@ Return true if `action_profile` is Pareto dominant for game `g`.
...
+        D = reshape(D, (size(D)[1], prod(size(D)[2:end])))
+    end
+
+    m, n = size(D)
+
+    A = Array{Float64}(undef, n+2, m+1)
+    A[1:n, 1:m] = -transpose(D)
+    A[1:n, end] .= 1.
+    A[n+1, 1:m] .= 1.
+    A[n+2, 1:m] .= -1.
```
Comment (oyamad) [inline src/normal_form_game.jl]: I believe you can pass both inequality and equality constraints to `linprog`. (The corresponding Python code is a bit stupid.)

File: test/test_normal_form_game.jl [hunk 2]
```diff
@@ -141,6 +141,8 @@
         @test @inferred(payoff_vector(player, nothing)) == [0, 1]
         @test @inferred is_best_response(player, 2, nothing)
         @test @inferred(best_response(player, nothing)) == 2
+        @test is_dominated(player, 1) == true
```
Comment (oyamad) [inline test/test_normal_form_game.jl]: Maybe you can just do

```suggestion
        @test is_dominated(player, 1)
```

File: test/test_normal_form_game.jl [hunk 3]
```diff
@@ -141,6 +141,8 @@
...
+        @test is_dominated(player, 2) == false
```
Comment (oyamad) [inline test/test_normal_form_game.jl]: Maybe

```suggestion
        @test !is_dominated(player, 2)
```

?

**Comment by oyamad** [comment]
Possible enhancements:

* [ ] Add `dominated_actions`; see https://github.com/QuantEcon/QuantEcon.py/pull/415/commits/0404f1a0220590a4bced5c859a823330d201660f.
* [ ] Can you make the code more general so that if the input game is of `Rational` type, then the computation is done with exact arithmetic, if the input `lp_solver` supports it? (Cf. #54)

File: src/normal_form_game.jl [hunk 4]
```diff
@@ -691,3 +691,75 @@ Return true if `action_profile` is Pareto dominant for game `g`.
...
+"""
+function is_dominated(player::Player, action::PureAction;
+                      tol::Float64=1e-8,
+                      lp_solver::MathProgBase.AbstractMathProgSolver=
+                      ClpSolver())
+    payoff_array = player.payoff_array
+
+    if num_opponents(player) == 0
+        return maximum(payoff_array) > payoff_array[action] + tol
+    end
```
Comment (oyamad) [inline src/normal_form_game.jl]: Should use multiple dispatch. (See for example `payoff_vector`.)

**Comment by QBatista** [comment]
@oyamad Among the LP solvers in `MathProgBase`, it looks like the only one which supports exact arithmetic is GLPK (it's a little bit difficult to find information about this so I may be wrong). Besides, it looks like GLPK is converting the inputs internally to do exact arithmetic. For example:
```julia
>>> b = [1/7, 2/7, 4/7, 1/4, 1/4, 1/2]
>>> c = [0., 1., 1., 1., 0, 1., 1., 1., 0.] 
>>> A = [1. 1. 1. 0. 0. 0. 0. 0. 0.;
         0. 0. 0. 1. 1. 1. 0. 0. 0.;
         0. 0. 0. 0. 0. 0. 1. 1. 1.;
         1. 0. 0. 1. 0. 0. 1. 0. 0.;
         0. 1. 0. 0. 1. 0. 0. 1. 0.;
         0. 0. 1. 0. 0. 1. 0. 0. 1.]
>>> linprog(c, A, '=', b, GLPKSolverLP(method=:Exact)).objval - 3/28
0.0
````
and
```julia
>>> b = [1//7, 2//7, 4//7, 1//4, 1//4, 1//2]
>>> c = [0//1, 1//1, 1//1, 1//1, 0//1, 1//1, 1//1, 1//1, 0//1]
>>> A = [1//1 1//1 1//1 0//1 0//1 0//1 0//1 0//1 0//1;
         0//1 0//1 0//1 1//1 1//1 1//1 0//1 0//1 0//1;
         0//1 0//1 0//1 0//1 0//1 0//1 1//1 1//1 1//1;
         1//1 0//1 0//1 1//1 0//1 0//1 1//1 0//1 0//1;
         0//1 1//1 0//1 0//1 1//1 0//1 0//1 1//1 0//1;
         0//1 0//1 1//1 0//1 0//1 1//1 0//1 0//1 1//1]
>>> linprog(c, A, '=', b, GLPKSolverLP(method=:Exact)).objval - 3/28
0.0
```
both work. However, I am guessing that there are cases where the output would differ based on the input type since `1//3 == Rational(1/3)` is `false`.

**Comment by QBatista** [comment]
This should now be ready.

File: src/normal_form_game.jl [hunk 5]
```diff
@@ -691,3 +691,112 @@ Return true if `action_profile` is Pareto dominant for game `g`.
...
+function is_dominated(player::Player, action::PureAction;
+                      tol::Float64=1e-8,
+                      lp_solver::MathProgBase.AbstractMathProgSolver=
+                      ClpSolver())
+    payoff_array = player.payoff_array
+    if eltype(player.payoff_array) == Rational
+        input_game_type = Rational
+    else
+        input_game_type = Float64
+    end
```
Comment (oyamad) [inline src/normal_form_game.jl]: Follow [this approach](https://github.com/QuantEcon/Games.jl/blob/34c99fc875cf97eccf965146217c9d796560110b/src/support_enumeration.jl#L99) in `support_enumeration`. Note that `Player` has a type parameter `T`.

File: src/Games.jl [hunk 6]
```diff
@@ -6,6 +6,7 @@ using LinearAlgebra, Random
 # Packages
 using Clp
 using MathProgBase
+using GLPKMathProgInterface
```
Comment (oyamad) [inline src/Games.jl]: Is this used in `src/normal_form_game.jl`?

File: test/test_normal_form_game.jl [hunk 7]
```diff
@@ -260,6 +284,59 @@
...
+            @test !is_dominated(player, action, tol=e+1e-16)
+            @test dominated_actions(player, tol=e+1e-16) ==
+                  Vector{Integer}(undef, 0)
+            @test is_dominated(player, action, tol=e/2)
+            @test dominated_actions(player, tol=e/2) == [action]
+        end
+
+        @testset "Test rational input game" begin
+            game_matrix = [2//3 1//3;
+                           1//3 2//3]
```
Comment (oyamad) [inline test/test_normal_form_game.jl]: Please test with a "corner case". You may use the above case with, e.g., `e` = `0//1` and `1//(10^8)`.

File: src/normal_form_game.jl [hunk 8]
```diff
@@ -691,3 +691,112 @@ Return true if `action_profile` is Pareto dominant for game `g`.
...
+  ClpSolver().
+
+# Returns
+
+- `::Bool` : True if `action` is strictly dominated by some mixed action;
+  False otherwise.
+
+"""
+function is_dominated(player::Player, action::PureAction;
+                      tol::Float64=1e-8,
```
Comment (oyamad) [inline src/normal_form_game.jl]:
```suggestion
                      tol::Real=1e-8,
```

**Comment by QBatista** [comment]
@oyamad I've made the changes. One counterintuitive thing to note is that `linprog` returns `Float64` even when `GLPKSolverLP(method=:Exact)` is used. Therefore, the tolerance test is actually not done using exact arithmetic since rationals get promoted to floats.

**Comment by oyamad** [comment]
Then will you try `CDDSolver(exact=true)` from [CDDLib.jl](https://github.com/JuliaPolyhedra/CDDLib.jl)?

File: src/normal_form_game.jl [hunk 9]
```diff
@@ -691,3 +691,108 @@ Return true if `action_profile` is Pareto dominant for game `g`.
...
+- `::Bool` : True if `action` is strictly dominated by some mixed action;
+  False otherwise.
+
+"""
+function is_dominated(player::Player{N,T}, action::PureAction;
+                      tol::Real=1e-8,
+                      lp_solver::MathProgBase.AbstractMathProgSolver=
+                      ClpSolver()) where {N,T<:Real}
+    payoff_array = player.payoff_array
+    input_game_type = typeof(zero(T)/one(T))
```
Comment (oyamad) [inline src/normal_form_game.jl]: I would use `S` for example, rather than `input_game_type`.

File: src/normal_form_game.jl [hunk 10]
```diff
@@ -691,3 +691,108 @@ Return true if `action_profile` is Pareto dominant for game `g`.
...
+# Returns
+
+- `out::Vector{Integer}` : Vector of integers representing pure actions, each
+  of which is strictly dominated by some mixed action.
+
+"""
+function dominated_actions(player::Player; tol::Real=1e-8,
+                           lp_solver::MathProgBase.AbstractMathProgSolver=
+                           ClpSolver())
+    out = Vector{Integer}(undef, 0)
```
Comment (oyamad) [inline src/normal_form_game.jl]: Do not use abstract type:

```suggestion
    out = Vector{Int}(undef, 0)
```

File: test/test_normal_form_game.jl [hunk 11]
```diff
@@ -260,6 +284,69 @@
...
         end
+
+        @testset "Test is_dominated" begin
+            coordination_game_matrix = [4 0; 3 2]
+            player = Player(coordination_game_matrix)
+            for action = 1:num_actions(player)
+                @test !is_dominated(player, action)
+            end
+
+            payoffs_2opponents = Array{Int64}(undef, 2, 2, 2)
```
Comment (oyamad) [inline test/test_normal_form_game.jl]:
```suggestion
            payoffs_2opponents = Array{Int}(undef, 2, 2, 2)
```

File: src/normal_form_game.jl [hunk 12]
```diff
@@ -691,3 +691,108 @@ Return true if `action_profile` is Pareto dominant for game `g`.
...
+    m, n = size(payoff_array, 1) - 1, prod(size(player.payoff_array)[2:end])
+
+    ind = trues(num_actions(player))
+    ind[action] = false
+
+    A = Array{input_game_type}(undef, n+1, m+1)
+    A[1:n, 1:m] = transpose(reshape(-selectdim(payoff_array, 1, ind) .+
+                                    selectdim(payoff_array, 1, action:action),
+                                    (m, n)))
+    A[1:n, m+1] .= 1.
```
Comment (oyamad) [inline src/normal_form_game.jl]:
```suggestion
    A[1:n, m+1] .= 1
```

File: src/normal_form_game.jl [hunk 13]
```diff
@@ -691,3 +691,108 @@ Return true if `action_profile` is Pareto dominant for game `g`.
...
+    A[n+1, 1:m] .= 1.
+    A[n+1, m+1] = 0.
```
Comment (oyamad) [inline src/normal_form_game.jl]:
```suggestion
    A[n+1, m+1] = 0
```

File: src/normal_form_game.jl [hunk 14]
```diff
@@ -691,3 +691,108 @@ Return true if `action_profile` is Pareto dominant for game `g`.
...
+
+    b = Array{input_game_type}(undef, n+1)
```
Comment (oyamad) [inline src/normal_form_game.jl]:
```suggestion
    b = zeros(S, n+1)
    b[end] = 1
```

File: test/test_normal_form_game.jl [hunk 15]
```diff
@@ -30,6 +34,9 @@
         # Perturbed best response
         @test best_response(player, [2/3, 1/3], [0., 0.1]) == 2
         @test best_response(player, [2, 1], [0., 0.1]) == 2
+
+        # Dominated actions
+        @test dominated_actions(player) == Vector{Integer}(undef, 0)
```
Comment (oyamad) [inline test/test_normal_form_game.jl]: Maybe
```suggestion
        @test dominated_actions(player) == Int[]
```
?

File: test/test_normal_form_game.jl (hunk 11 above)
Comment (QBatista) [reply test/test_normal_form_game.jl]: There are multiple existing tests that use `Int64` -- should this change be made there as well?

**Comment by QBatista** [comment]
@oyamad I've made the changes -- please let me know if there are any additional ones that you want me to make.

File: src/normal_form_game.jl [hunk 16]
```diff
@@ -691,3 +691,107 @@ Return true if `action_profile` is Pareto dominant for game `g`.
...
+                      lp_solver::MathProgBase.AbstractMathProgSolver=
+                      ClpSolver())
+        payoff_array = player.payoff_array
+        return maximum(payoff_array) > payoff_array[action] + tol
+end
+
+# dominated_actions
+
+"""
+    dominated_actions(player[, tol=1e-8])
```
Comment (oyamad) [inline src/normal_form_game.jl]: The argument `lp_solver` should also be added.

File: test/test_normal_form_game.jl [hunk 17]
```diff
@@ -260,6 +284,67 @@
...
+
+            e = 1 // 1
+            player = Player([-e -e;
+                             10^8//1 -10^8//1;
+                             -10^8//1 10^8//1])
+
+            action = 1
+            @test !is_dominated(player, action, tol=e,
+                                lp_solver=lp_solver)
+            @test is_dominated(player, action, tol=0//1, lp_solver=lp_solver)
```
[the remark this answers is not in the document]
Comment (QBatista) [reply test/test_normal_form_game.jl]: @oyamad Sorry, I forgot to explain this. These tests are meant to mimic the [Python ones](https://github.com/QuantEcon/QuantEcon.py/blob/01ee68dbbd815b6275192a707377bb12e5e9da8f/quantecon/game_theory/tests/test_normal_form_game.py#L165-L175). With the game matrix from the Python version (which is similar to yours) and  `e=1//10^8`, there is some approximation error and `res.sol[end]` is not exactly equal to `e` even when `CDDSolver(exact=true)`. When multiplying the game matrix by `10^8`, this approximation error disappears, that is, `res.sol[end]` equals `e`, which are both equal to `1`. Setting `tol=e` is a corner case in the sense that `1//1 > 1//1` is computed internally. 

That being said, I think your test case is easier to understand and more appropriate given that the function is intended to compute dominance, so I can get rid of my tests if you think they are not helpful.

File: REQUIRE [hunk 18]
```diff
@@ -4,4 +4,4 @@ MathProgBase
 QuantEcon
 Polyhedra
 LightGraphs
-Combinatorics
+Combinatorics
```
Comment (oyamad) [inline REQUIRE]: Do not remove the line break.

File: src/normal_form_game.jl [hunk 19]
```diff
@@ -691,3 +691,107 @@ Return true if `action_profile` is Pareto dominant for game `g`.
...
+    A[n+1, m+1] = 0
+
+    b = zeros(S, n+1)
+    b[end] = 1
+
+    c = zeros(S, m+1)
+    c[end] = -1
+
+    sense = Array{Char}(undef, n+1)
+    sense[1:n] .= '<'
```
Comment (oyamad) [inline src/normal_form_game.jl]: Here I got

```
┌ Warning: broadcast will default to iterating over its arguments in the future. Wrap arguments of
│ type `x::Char` with `Ref(x)` to ensure they broadcast as "scalar" elements.
│   caller = ip:0x0
└ @ Core :-1
```

The warning says you should do

```jl
sense[1:n] .= Ref('<')
```

though I don't know what `Ref` means.

Alternatively you can just write

```jl
for i in 1:n
    sense[i] = '<'
end
```

File: test/test_normal_form_game.jl (hunk 17 above)
[the remark this answers is not in the document]
Comment (oyamad) [reply test/test_normal_form_game.jl]: As a workaround, let `e` be one divided by a multiple of `2`, for example `1//(2^25)`, in the instance in my previous comment.

I opened an issue in JuliaPolyhedra/CDDLib.jl#31.

File: src/normal_form_game.jl [hunk 20]
```diff
@@ -691,3 +691,107 @@ Return true if `action_profile` is Pareto dominant for game `g`.
...
+
+"""
+    dominated_actions(player[, tol=1e-8, lp_solver=ClpSolver()])
+
+Return a vector of actions that are strictly dominated by some mixed actions.
+
+# Arguments
+
+- `player::Player` : Player instance.
+- `tol::Real` : Tolerance level used in determining domination.
```
Comment (oyamad) [inline src/normal_form_game.jl]: Add a description of `lp_solver`.

```
- `lp_solver::AbstractMathProgSolver` : See `is_dominated`.
```
